scripts/lint_mathlib.lean

```diff
@@ -0,0 +1,91 @@
+/-
+Copyright (c) 2020 Robert Y. Lewis. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Robert Y. Lewis, Gabriel Ebner
+-/
+
+import tactic.lint system.io  -- these are required
+import all   -- then import everything, to parse the library for failing linters
+
+/-!
+# lint_mathlib
+
+Script that runs the linters listed in `mathlib_linters` on all of mathlib.
+As a side effect, the file `nolints.txt` is generated in the current
+directory. This script needs to be run in the root directory of mathlib.
+
+It assumes that files generated by `mk_all.sh` are present.
+
+This is used by the CI script for mathlib.
+
+Usage: `lean --run scripts/lint_mathlib.lean`
+-/
+
+open native
+
+/--
+Returns the contents of the `nolints.txt` file.
+-/
+meta def mk_nolint_file (env : environment) (mathlib_path_len : ℕ)
+  (results : list (name × linter × rb_map name string)) : format := do
+let failed_decls_by_file := rb_lmap.of_list (do
+  (linter_name, _, decls) ← results,
+  (decl_name, _) ← decls.to_list,
+  let file_name := (env.decl_olean decl_name).get_or_else "",
+  pure (file_name.popn mathlib_path_len, decl_name.to_string, linter_name.last)),
+format.intercalate format.line $
+"import .all" ::
+"run_cmd tactic.skip" :: do
+(file_name, decls) ← failed_decls_by_file.to_list.reverse,
+"" :: ("-- " ++ file_name) :: do
+(decl, linters) ← (rb_lmap.of_list decls).to_list.reverse,
+pure $ "apply_nolint " ++ decl ++ " " ++ " ".intercalate linters
+
+/--
+Parses the list of lines of the `nolints.txt` into an `rb_lmap` from linters to declarations.
+-/
+meta def parse_nolints (lines : list string) : rb_lmap name name :=
+rb_lmap.of_list $ do
+line ← lines,
+guard $ line.front = 'a',
+_ :: decl :: linters ← pure $ line.split (= ' ') | [],
+let decl := name.from_components $ decl.split (= '.'),
```

Comment:
Do we already have a function that parses a name into a string (i.e., ``parse_name "foo.bar.1" = `foo.bar.1``)?

Reply:
Actually, I don't think we do. This would be a good addition.